storage-proofs/src/rational_post.rs

```diff
@@ -0,0 +1,473 @@
+    type PublicParams = PublicParams;
+    type SetupParams = SetupParams;
+    type PublicInputs = PublicInputs<'a, H::Domain>;
+    type PrivateInputs = PrivateInputs<'a, H>;
+    type Proof = Proof<H>;
+    type Requirements = NoRequirements;
+
+    fn setup(sp: &Self::SetupParams) -> Result<Self::PublicParams> {
+        Ok(PublicParams {
+            sector_size: sp.sector_size,
+            challenges_count: sp.challenges_count,
+        })
+    }
+
+    fn prove<'b>(
+        _pub_params: &'b Self::PublicParams,
+        pub_inputs: &'b Self::PublicInputs,
+        priv_inputs: &'b Self::PrivateInputs,
+    ) -> Result<Self::Proof> {
+        assert_eq!(
+            pub_inputs.challenges.len(),
+            pub_inputs.commitments.len(),
+            "mismatched challenges and commitments"
+        );
+        let challenges = pub_inputs.challenges;
+
+        let proofs = challenges
+            .iter()
+            .zip(pub_inputs.commitments.iter())
+            .map(|(challenge, commitment)| {
+                let challenged_sector = challenge.sector;
+                let challenged_leaf = challenge.leaf;
+
+                let tree = priv_inputs.trees[challenged_sector as usize];
+                if commitment != &tree.root() {
+                    return Err(Error::InvalidCommitment);
+                }
+
+                Ok(MerkleProof::new_from_proof(
+                    &tree.gen_proof(challenged_leaf as usize),
+                ))
+            })
+            .collect::<Result<Vec<_>>>()?;
+
+        Ok(Proof {
+            inclusion_proofs: proofs,
+        })
+    }
+
+    fn verify(
+        pub_params: &Self::PublicParams,
+        pub_inputs: &Self::PublicInputs,
+        proof: &Self::Proof,
+    ) -> Result<bool> {
+        let challenges = pub_inputs.challenges;
+
+        if challenges.len() != pub_inputs.commitments.len() as usize {
+            return Err(Error::MalformedInput);
+        }
+
+        if challenges.len() != proof.inclusion_proofs.len() {
+            return Err(Error::MalformedInput);
+        }
+
+        // validate each proof
+        for ((merkle_proof, challenge), commitment) in proof
+            .inclusion_proofs
+            .iter()
+            .zip(challenges.iter())
+            .zip(pub_inputs.commitments.iter())
+        {
+            let challenged_leaf = challenge.leaf;
+
+            // validate the commitment
+            if merkle_proof.root() != commitment {
+                return Ok(false);
+            }
+
+            // validate the path length
+            if graph_height(pub_params.sector_size as usize / NODE_SIZE)
+                != merkle_proof.path().len()
+            {
+                return Ok(false);
+            }
+
+            if !merkle_proof.validate(challenged_leaf as usize) {
+                return Ok(false);
+            }
+        }
+
+        Ok(true)
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Challenge {
+    pub sector: u64,
```

Comment:
BLOCKING: Is this a sector id? An index of some sort? You either need a better name for this field, a better type than `u64`, or some comments. Or maybe all three.

Reply:
It is a sector id, and it has the same name as in the spec

Reply:
> It is a sector id

OK. Would you change the struct field from `sector` to `sector_id`?

Reply:
no, I would rather stay in sync with the spec

Reply:
> no, I would rather stay in sync with the spec

Are you referring [to the field names used in the Challenge Derivation](https://github.com/filecoin-project/specs/blob/master/proof-of-spacetime.md#challenge-derivation) _pseudocode_?

Reply:
I'm a bit confused by this. If it is indeed a sector id, we should probably use that term in this part of the spec — since it appears elsewhere in the spec. (But this is a spec question, and spec/implementation don't *have* to agree.)

I wonder, though, if it is actually a sector id at all? Doesn't challenge derivation produce an index into an array of sectors? As far as I know, the set of sectors being challenged won't usually have sequential sector ids, starting from 0 — but I do think challenge derivation assumes that the `sector` value produced will index into such a set of values. Since you've decided not to use `sector_id` anyway, this is mostly just a clarification question to make sure we're all on the same page about what is happening.

It might also mean this line of the spec needs changing:
>         attemptedSectors := {SectorID:bool}

Reply:
> As far as I know, the set of sectors being challenged won't usually have sequential sector ids, starting from 0 — but I do think challenge derivation assumes that the sector value produced will index into such a set of values.

Reviewing the `GeneratePoSt` pseudocode:

```
func GeneratePoSt(
    sectorSize BytesAmount, 
    sectors SectorSet, 
    seed Seed, 
    faults FaultSet
) PoStProof {
```

An important thing to note here is that `SectorSet` is a sector id-indexed data structure (effectively a map from sector id to bool). It is not something you can index in to by _offset_.

In the guts of `GeneratePoSt`, after generating the inclusion proof, we add sector ids (from `sectors`) into `sectorsSorted`. We can index into a `SectorSet` by sector id only - which implies that `challenge.Sector` is a sector id:

```
sectorsSorted[i] = sectors[challenge.Sector]
```

Reviewing the challenge _generation_ pseudocode, it becomes clear that the `Sector` field's value is intended to be an _offset_ in the `SectorSet`:

```
    return Challenge {
        Sector: sector_challenge % sectorCount,
        Leaf: leaf_challenge % (sectorSize / NODE_SIZE),
    }
```

We either need to change, in the spec, the definition of `SectorSet` such that it becomes a sorted list of sector ids (into which we can index by offset) OR we need to rethink challenge generation.

Reply:
I don't think challenge generation itself should be modified. There's no sane world in which it produces sector ids. I don't mind which option for fussing with the plumbing we use, though.